src/uuid/mod.rs

```diff
@@ -0,0 +1,461 @@
+//! #   /// on whether `bson::to_bson` or `bson::to_vec` is used.
+//! #   uuid: uuid::Uuid,
+//! #
+//! #   /// serializes as a BSON binary with subtype 4.
+//! #   bson_uuid: bson::Uuid,
+//! # }
+//! use serde_json::json;
+//!
+//! let uuid = uuid::Uuid::new_v4();
+//! let bson_uuid: bson::Uuid = uuid.into();
+//! let foo = Foo { uuid, bson_uuid, };
+//!
+//! let json = serde_json::to_value(&foo)?;
+//! assert_eq!(json, json!({ "uuid": uuid.to_string(), "bson_uuid": uuid.to_string() }));
+//! # }
+//! # Ok::<(), Box::<dyn std::error::Error>>(())
+//! ```
+#[cfg(test)]
+mod test;
+
+use std::{
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{de::BsonVisitor, spec::BinarySubtype, Binary, Bson};
+
+pub(crate) const UUID_NEWTYPE_NAME: &str = "BsonUuid";
+
+/// A struct modeling a BSON UUID value (i.e. a Binary value with subtype 4).
+///
+/// This type should be used instead of [`uuid::Uuid`](https://docs.rs/uuid/latest/uuid/struct.Uuid.html)
+/// when serializing to or deserializing from BSON, since
+/// [`uuid::Uuid`](https://docs.rs/uuid/latest/uuid/struct.Uuid.html)'s `serde` implementation doesn't
+/// produce or parse BSON UUIDs.
+///
+/// To enable interop with the `Uuid` type from the `uuid` crate, enable the `uuid-0_8` feature
+/// flag.
+///
+/// For more information on the usage of this type, see the [`uuid`] module-level documentation.
+#[derive(Clone, Copy, PartialEq, Hash)]
+pub struct Uuid {
+    uuid: uuid::Uuid,
+}
+
+impl Uuid {
+    /// Creates a random UUID.
+    ///
+    /// This uses the operating system's RNG as the source of random numbers. If you'd like to use a
+    /// custom generator, generate random bytes and pass them to [`Uuid::from_bytes`] instead.
+    pub fn new() -> Self {
+        Self {
+            uuid: uuid::Uuid::new_v4(),
+        }
+    }
+
+    /// Creates a [`Uuid`] using the supplied big-endian bytes.
+    pub const fn from_bytes(bytes: [u8; 16]) -> Self {
+        Self::from_external_uuid(uuid::Uuid::from_bytes(bytes))
+    }
+
+    /// Creates a [`Uuid`] from the provided hex string.
+    pub fn parse_str(input: impl AsRef<str>) -> Result<Self> {
+        let uuid = uuid::Uuid::parse_str(input.as_ref()).map_err(|e| Error::InvalidUuidString {
+            message: e.to_string(),
+        })?;
+        Ok(Self::from_external_uuid(uuid))
+    }
+
+    pub(crate) const fn from_external_uuid(uuid: uuid::Uuid) -> Self {
+        Self { uuid }
+    }
+
+    /// Returns an array of 16 bytes containing the [`Uuid`]'s data.
+    pub const fn bytes(self) -> [u8; 16] {
+        *self.uuid.as_bytes()
+    }
+}
+
+impl Default for Uuid {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(feature = "uuid-0_8")]
+#[cfg_attr(docsrs, doc(cfg(feature = "uuid-0_8")))]
+impl Uuid {
+    /// Convert this [`Uuid`] to a [`uuid::Uuid`] from the [`uuid`](https://docs.rs/uuid/latest) crate.
```

Comment:
I figured it would be confusing to call this `from_uuid`, so I just appended the version number to the end to disambiguate. This will also be useful for when `uuid` releases another minor version.